datafusion/core/src/physical_optimizer/sort_pushdown.rs

```diff
@@ -103,7 +110,12 @@ fn pushdown_sorts_helper(
         if !satisfy_parent {
             // Make sure this `SortExec` satisfies parent requirements:
             let sort_reqs = requirements.data.ordering_requirement.unwrap_or_default();
-            let fetch = requirements.data.fetch;
+            let fetch = if let Some(fetch) = requirements.data.fetch.clone() {
+                Some(fetch)
+            } else {
+                // It's possible current plan (`SortExec`) has a fetch value.
+                sort_fetch
+            };
```

Comment:
I think you can express this same thing more concisely
```suggestion
            let fetch =  requirements.data.fetch.clone()
                // It's possible current plan (`SortExec`) has a fetch value.
                .and(sort_fetch);
```

Also it seems strange that we have two fetches here that could be different . 

I reverted just this change and the test still passes. By my reading of the code you have fixed the issue above by setting the requirements `fetch`.

So maybe this would be better as check that the fetches are the same and an internal error if they aren't 🤔

Reply:
Yes, they should be the same, checking is enough.

For the suggestion, I think what you mean is `requirements.data.fetch.clone().or_else(|| sort_fetch)`?

Reply:
There is a test which has different values
```rust
 async fn test_remove_unnecessary_sort8() -> Result<()> {
        let schema = create_test_schema()?;
        let source = memory_exec(&schema);
        let input = Arc::new(SortExec::new(
            LexOrdering::new(vec![sort_expr("non_nullable_col", &schema)]),
            source,
        ));
        let limit = Arc::new(LocalLimitExec::new(input, 2));
        let physical_plan = sort_exec(
            vec![
                sort_expr("non_nullable_col", &schema),
                sort_expr("nullable_col", &schema),
            ],
            limit,
        );

        let expected_input = [
            "SortExec: expr=[non_nullable_col@1 ASC, nullable_col@0 ASC], preserve_partitioning=[false]",
            "  LocalLimitExec: fetch=2",
            "    SortExec: expr=[non_nullable_col@1 ASC], preserve_partitioning=[false]",
            "      MemoryExec: partitions=1, partition_sizes=[0]",
        ];
        let expected_optimized = [
            "LocalLimitExec: fetch=2",
            "  SortExec: TopK(fetch=2), expr=[non_nullable_col@1 ASC, nullable_col@0 ASC], preserve_partitioning=[false]",
            "    MemoryExec: partitions=1, partition_sizes=[0]",
        ];
        assert_optimized!(expected_input, expected_optimized, physical_plan, true);

        Ok(())
    }
```
requirement data fetch is Some(2), but sort doesn't have fetch, is none.